rfd/0157-aws-eks-discover.md

```diff
@@ -0,0 +1,205 @@
+allowing us to avoid the need for an additional step to run Helm itself elsewhere. It would require
+running a job inside the target EKS cluster with a special image that has Helm on it, and this job
+would install the Teleport kube agent. By using the Helm Go SDK, we can run the installation directly from the Teleport
+process and don't need to maintain a new special image and EKS installation job.
+
+### Automatic Discovery and Enrollment
+
+Users will also be able to set up automatic discovery and enrollment of EKS clusters in the Discover UI.
+We will reuse the existing EKS discovery capability, although a new mode will be added to it - discovery through
+the AWS integration. EKS discovery in that mode will rely on the AWS integration, selected in the Discovery UI,
+to find clusters and then install the Helm chart on them. The algorithm will be as follows:
+- List EKS clusters through the AWS integration.
+- Cross-check to see which EKS clusters are already registered in the system.
+- Once a new EKS cluster is found, associate our OIDC provider with it.
+- Using the associated provider, generate an access token and install the kube agent through the Helm chart.
+
+Users will be able to filter EKS clusters for discovery by using labels.
+
+A dynamic discovery configuration will be created as a result of users setting up automatic EKS discovery in the Discover UI.
+This configuration can be picked up by any discovery service. In the Teleport cloud, the Discovery service will be available by default,
+meaning users will not need to perform any additional actions.
+
+### EKS Identity Provider Association
+
+AWS EKS supports a method of authentication that involves associating an OIDC identity provider with the cluster.
+Such an associated identity provider can issue access tokens (JWT) that grant access to the EKS cluster groups and users.
+Since Teleport already supports OIDC providers, we can easily use this approach. It will require slight
+modifications to the AWS OIDC integration code to allow it to issue JWTs for an EKS cluster. To associate an OIDC provider with an
+EKS cluster, the selected AWS integration needs to issue an API call `eks:AssociateIdentityProviderConfig`
+with the following configuration (includes example parameters):
+
+```yaml
+---
+apiVersion: eksctl.io/v1alpha5
+kind: ClusterConfig
+
+metadata:
+  name: eks-cluster-name
+  region: us-east-1
+
+identityProviders:
+  - name: TeleportProvider
+    type: oidc
+    issuerUrl: https://teleport.example.com
+    clientId: kubernetes
+    usernameClaim: teleport-eks-oidc-user
+    groupsClaim: teleport-eks-oidc-groups
+```
+
+To access the target EKS cluster in API mode, we will generate an access token using the associated OIDC provider.
+That token will grant access to the `system:masters` group, so the Helm chart installation will be able to set up
+all the required infrastructure for the Teleport kube agent, including service accounts.
```

Comment:
How do you select the scopes for the generated access token? I imagine some users might not want to give Teleport system:masters access and instead have a pre-configured service account (or cluster role), say scoped down to a particular namespace where they want Teleport to run. Is there a way we can provide an option for them to do that somehow?

Reply:
We can make it configurable, storing user/group claims we should use on the integration, but we'll need some UI for this. Or we don't store it on the integration, but you'll need to specify user/group claims at the moment you want to enrol selected clusters or setup auto discovery (which also will require some ui, maybe a bit easier that with integration, but you'll need to input it each time).

Reply:
Added this to the Future Considerations section.